### PR TITLE
[choreo] Document Updater

### LINKED ISSUE #570: Reimplement data upgrader in Rust for both types of 2025 files

### PR DESCRIPTION
**Pull request by oh-yes-0-fps**
Needs unit tests but has been tested on my robot project

This pr also adds `trackwidth` to .traj output field to show off the upgrader, its just a dummy value of 1.0 for now

Fixes #570.

### CODE REVIEW

**Comment by calcmogul** [comment]
What are the steps for adding another schema version upgrade?

**Comment by oh-yes-0-fps** [comment]
> What are the steps for adding another schema version upgrade?

Incrementing the constants for the schema version (its just an incrementing integer now for simplicity idk how you define semver for the schema stuff when any change is "breaking").
Add a new upgrader action in `upgraders.rs` for the changes you made
Update the choreolib code if its needed.

File: src-core/src/spec/trajectory.rs [hunk 1]
```diff
@@ -399,6 +399,9 @@ pub struct Trajectory {
     /// This includes 0, but the index of the last sample is never in this list even if the split toggle is set
     /// for the last waypoint
     pub splits: Vec<usize>,
+    /// The trackwidth of the robot,
+    /// used in robot code for differential drive calculations
+    pub trackwidth: f64,
```
Comment (shueja) [inline src-core/src/spec/trajectory.rs:404]: Either this should be optional or we have to define it for swerve

**Review by shueja — CHANGES_REQUESTED** [review]
Architecturally this seems reasonable. I'll have to think about ramifications of changing save file versioning to an int. I also don't know if we should require that Choreolib changes matching spec changes must go in the same PR.

File: src-core/src/spec/trajectory.rs (hunk 1 above)
Comment (oh-yes-0-fps) [reply src-core/src/spec/trajectory.rs:404]: Yeah, requiring it for swerve isnt bad, the data is already in robot config

Comment (shueja) [reply src-core/src/spec/trajectory.rs:404]: Is it? Which distance between modules is used?

**Comment by spacey-sooty** [comment]
>  I also don't know if we should require that Choreolib changes matching spec changes must go in the same PR.

I think we risk creating issues/having Choreolib fall behind similar to how C++ does if it's not updated in the same PR.

**Comment by shueja** [comment]
I agree. I'm just wondering what effect that will have on PRs getting carried through to mergeability.

**Comment by calcmogul** [comment]
If we can't get it done in the PR itself, I'd be concerned about it ever getting done.

**Comment by calcmogul** [comment]
Items to address:

- [ ] What is the "trackwidth" field when it's a swerve trajectory
- [ ] ChoreoLib C++/Java/Python changes to use an int for version number (and fail non-fatally if it's a string)
- [ ] Unit tests for the already implemented upgrader, and a clear place to put more

Item 3 is somewhat related to another issue about assembling a common set of test JSONs for all sorts of tests.

File: src-core/src/spec/trajectory.rs (hunk 1 above)
Comment (oh-yes-0-fps) [reply src-core/src/spec/trajectory.rs:404]: between front left and front right ig, we can document that behavior

**Comment by shueja** [comment]
Trackwidth work moved to https://github.com/SleipnirGroup/Choreo/issues/969

**Review by calcmogul — APPROVED** [review]

**Pull request merged**